File: 2015/Python/day01.py
def part_one(string: str) -> int:
    floor = 0
    for char in string:
        if char == '(': floor += 1
        else: floor -= 1
    return floor

def part_two(string: str) -> int:
    floor = 0
    pos = 0
    for char in string:
        pos += 1
        if char == '(': floor += 1
        else: floor -= 1
        if floor == -1:
            return pos;

with open("../data/day01.txt", "r") as file:
    brackets = file.read()
    print(part_one(brackets))
    print(part_two(brackets))

File: 2015/Scala/day01.scala
import scala.io.Source
import scala.util.Using

object day01 {

  def partOne(directions: List[Char]): Int = {
    directions.map {
      case '(' => 1
      case ')' => -1
    }.sum
  }

  def partTwo(directions: List[Char]): Int = {
    directions.map {
      case '(' => 1
      case ')' => -1
    }.scanLeft(0)(_ + _).indexWhere(_ < 0)
  }

  def main(args: Array[String]): Unit = {
    val inputFile = "2015/data/day01.txt"
    val brackets = Using.resource(Source.fromFile(inputFile)) { source =>
      source.mkString.toList
    }

    println(partOne(brackets))
    println(partTwo(brackets))
  }
}
